### ISSUE TITLE
AssignmentInConditionSniff false positive in for() loop using comma operator

### ISSUE THREAD

**Issue by gnutix** [squizlabs/PHP_CodeSniffer#3418]
**Describe the bug**
When using the comma operator in a for() loop, PHP_CodeSniffer reports an error in the AssignmentInConditionSniff.

**Code sample**

Here a complete example of such a loop :

```php
/**
 * @return T|null
 */
public function valueAt(LocalDateTime $time): mixed
{
    $items = $this->items;

    /** @noinspection SuspiciousLoopInspection use of the "comma operator" is a rare syntax only for `for()` */
    for (
        $start = 0, $end = count($items);
        $i = $start + (int) (($end - $start) / 2), $i < $end;
    ) {
        /**
         * @var LocalDateTimeInterval $itemTimeRange
         * @var T $value
         */
        [$itemTimeRange, $value] = $items[$i];
        if ($itemTimeRange->contains($time)) {
            return $value;
        } elseif ($itemTimeRange->isBefore($time)) {
            $start = $i + 1;
        } else {
            $end = $i;
        }
    }

    return null;
}
```

**To reproduce**

I'm running PHP_CodeSniffer through ECS using `SetList::COMMON` (which includes AssignmentInConditionSniff as part of the `control-structures.php` file), which produces the following output : 

```
 ------------------------------------------------------------------------------------------------------------------------------------
 src/Object.php:437
 ------------------------------------------------------------------------------------------------------------------------------------
 Variable assignment found within a condition. Did you mean to do a comparison ?

 Reported by: "Found"
 ------------------------------------------------------------------------------------------------------------------------------------
```

**Expected behavior**
It should not report this code as an assignment in a condition, as it is not.

**Versions (please complete the following information):**
 - OS: Linux
 - PHP: 8.0.7
 - ECS: 9.4.40

**Additional context**
None.

**Comment by jrfnl** [comment]
I'm sorry, but I don't understand why you are saying that this is a false positive.

In the above example, the _condition_ part of the `for` statement is `$i = $start + (int) (($end - $start) / 2), $i < $end` and the sniff is IMO correctly being triggered for the `$i = ...` in that condition.

**Comment by gnutix** [comment]
Technically, it is maybe considered as a _condition_. Yet what I mean by _false positive_ is that there is no other way to write this `for` loop than by assigning the `$i` variable in the second part of the `for` loop (`expr2`, as referenced in the [doc](https://www.php.net/manual/en/control-structures.for.php)). I cannot move the `$i = ...` part in the first expression - it would simply not work. So this Sniff is incompatible with this usage of the `for` loop, which is a valid one programmatically speaking.

**Comment by jrfnl** [comment]
> there is no other way to write this `for` loop than by assigning the `$i` variable in the second part of the `for` loop

@gnutix Well, there is: you could move the assignment to be the last statement in the body of the `for`.

> And maybe that's fine, and I should just ignore it in my project. That would be a correct response for you guys. I just wanted to make sure that's how you feel about this use case you might not even have ever heard about (I had never seen such a loop before my colleague wrote it).

The sniff does exactly what is advertised: it flags all "assignments in conditions" and yes, while I'm perfectly aware this is valid PHP, similar to `while (($i = do_something()) === false) {...}`, it is an _assignment in condition_ and therefore will be flagged.

Assignments in conditions are generally considered _risky_. They are often a typo (comparison was intended) and can always be written in a way that the assignment does not have to be in the condition, so this sniff is specifically intended to find these risks/typos.

P.S.: No "guys" here.

**Comment by gnutix** [comment]
Sorry for my unintended use of non-inclusive language. 🙏

I will try adding the assignment in the first part of the for loop ASAP.

**Comment by gnutix** [comment]
I couldn't find any way to make it work without having that assignment. I give up, let's forget it. Thanks for your answers.

**Comment by jrfnl** [comment]
@gnutix Did you try what I suggested ?

```php
    for (
        $start = 0, $end = count($items);
        $i < $end;
    ) {
        /**
         * @var LocalDateTimeInterval $itemTimeRange
         * @var T $value
         */
        [$itemTimeRange, $value] = $items[$i];
        if ($itemTimeRange->contains($time)) {
            return $value;
        } elseif ($itemTimeRange->isBefore($time)) {
            $start = $i + 1;
        } else {
            $end = $i;
        }
        
        $i = $start + (int) (($end - $start) / 2)
    }
```

**Comment by gnutix** [comment]
Your example yields an `undefined variable $i`. The only way that works is copy/pasting the `$i = ...` part like so : 

```php
        for (
            $start = 0, $end = count($items), $i = $start + (int) (($end - $start) / 2);
            $i < $end;
        ) {
            /**
             * @var LocalDateTimeInterval $itemTimeRange
             */
            [$itemTimeRange,] = $items[$i];
            if ($itemTimeRange->intersects($timeRange)) {
                throw new TimelineRangeConflictException($timeRange, $itemTimeRange);
            }

            if ($itemTimeRange->isBeforeInterval($timeRange)) {
                // If $items[$i] is before the range, we drop the lower half of indices
                $start = $i + 1;
            } else {
                // If $items[$i] is after the range, we instead drop the higher half of indices
                $end = $i;
            }

            $i = $start + (int) (($end - $start) / 2);
        }
```

But then I find it much better to just have it written once, even though it means I have to ignore this sniff for this file.

**Comment by jrfnl** [comment]
As long as it works for you. And just for your info: you don't have to ignore the sniff for the whole file, you can use an inline ignore annotation to just ignore it for the one line: `//phpcs:ignore Generic.CodeAnalysis.AssignmentInCondition` (on the line above or as a trailing comment.
